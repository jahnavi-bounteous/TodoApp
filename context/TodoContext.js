import React, { createContext, useReducer, useContext, useEffect, useState } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';

export const ADD_TODO = 'ADD_TODO';
export const DELETE_TODO = 'DELETE_TODO';
export const EDIT_TODO = 'EDIT_TODO';
export const SET_TODOS = 'SET_TODOS';
export const TOGGLE_IMPT = 'TOGGLE_IMP';
export const MARK_COMPLETE = 'MARK_COMPLETE';
export const LOGIN_USER = 'LOGIN_USER';
export const LOGOUT_USER = 'LOGOUT_USER';

const TodoContext = createContext();

export const useTodoContext = () => {
  return useContext(TodoContext);
};

const todoReducer = (state, action) => {
  switch (action.type) {
    case SET_TODOS:
      return action.payload;
    case ADD_TODO:
      return [...state, action.payload];
    case DELETE_TODO:
      return state.filter((todo) => todo.id !== action.payload);
    case EDIT_TODO:
      return state.map((todo) =>
        todo.id === action.payload.id ? action.payload : todo
      );
    case TOGGLE_IMPT:
      return state.map((todo) =>
        todo.id === action.payload ? { ...todo, isImportant: !todo.isImportant } : todo
      );
    case MARK_COMPLETE:
      return state.map((todo) =>
        todo.id === action.payload ? { ...todo, isComplete: !todo.isComplete } : todo
      );
    default:
      return state;
  }
};

export const TodoProvider = ({ children }) => {
  const [todos, dispatch] = useReducer(todoReducer, []);
  const [loggedInUser, setLoggedInUser] = useState(null);

  useEffect(() => {
    const loadUserData = async () => {
      try {
        const userData = await AsyncStorage.getItem('loggedInUser');
        if (userData) {
          const { username } = JSON.parse(userData);
          setLoggedInUser(username);
          const storedTodos = await AsyncStorage.getItem(`todos_${username}`);
          if (storedTodos !== null) {
            dispatch({ type: SET_TODOS, payload: JSON.parse(storedTodos) });
          }
        }
      } catch (error) {
        console.error('Error loading user data from AsyncStorage:', error);
      }
    };
    loadUserData();
  }, []);

  useEffect(() => {
    const saveTodos = async () => {
      try {
        if (loggedInUser) {
          await AsyncStorage.setItem(`todos_${loggedInUser}`, JSON.stringify(todos));
        }
      } catch (error) {
        console.error('Error saving todos to AsyncStorage:', error);
      }
    };
    saveTodos();
  }, [todos, loggedInUser]);

  const loginUser = (username) => {
    setLoggedInUser(username);
  };

  const logoutUser = async () => {
    try {
      await AsyncStorage.removeItem('loggedInUser');
      setLoggedInUser(null);
    } catch (error) {
      console.error('Error logging out:', error);
    }
  };

  return (
    <TodoContext.Provider value={{ todos, dispatch, loggedInUser, loginUser, logoutUser }}>
      {children}
    </TodoContext.Provider>
  );
};