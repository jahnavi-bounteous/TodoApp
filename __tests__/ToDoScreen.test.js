import React from 'react';
import renderer from 'react-test-renderer';
import TodoScreen from '../screens/ToDoScreen';
import { TodoProvider } from '../context/TodoContext';

// Mock AsyncStorage
jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn(),
}));

// Mock Animated module
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

// Mock useNavigation hook
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({ navigate: jest.fn() }),
}));

describe('TodoScreen Component', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <TodoProvider>
          <TodoScreen />
        </TodoProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should handle search input', () => {
    const component = renderer.create(
      <TodoProvider>
        <TodoScreen />
      </TodoProvider>
    );
    const instance = component.root;

    const searchInput = instance.findByProps({ placeholder: 'Search Todos' });
    searchInput.props.onChangeText('Test');

    expect(instance.findByProps({ placeholder: 'Search Todos' }).props.value).toBe('Test');
  });

//   it('should navigate to AddTodo screen on button press', () => {
//     const component = renderer.create(
//       <TodoProvider>
//         <TodoScreen />
//       </TodoProvider>
//     );
//     const instance = component.root;

//     const addButton = instance.findByProps({ testID: 'addButton' });
//     addButton.props.onPress();

//     expect(instance.instance.props.children.props.navigation.navigate).toHaveBeenCalledWith(
//       'AddTodo'
//     );
//   });
});
