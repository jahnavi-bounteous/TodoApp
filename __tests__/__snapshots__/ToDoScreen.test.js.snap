// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoScreen Component renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "paddingHorizontal": 10,
      }
    }
  >
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Search Todos"
      style={
        [
          {
            "borderRadius": 5,
            "borderWidth": 2,
            "flex": 1,
            "marginRight": 1,
            "padding": 10,
          },
          {
            "borderColor": "#ccc",
          },
        ]
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "borderRadius": 20,
          "opacity": 1,
          "padding": 3,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderRadius": 23,
            "height": 46,
            "margin": 6,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
            "width": 46,
          }
        }
        testID="icon-button-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderColor": "rgba(121, 116, 126, 1)",
              "borderRadius": 23,
              "borderWidth": 0,
              "elevation": 0,
              "flex": 1,
              "overflow": "hidden",
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="icon-button-container"
        >
          <View
            accessibilityComponentType="button"
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityTraits="button"
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            centered={true}
            collapsable={false}
            focusable={true}
            hitSlop={
              {
                "bottom": 6,
                "left": 6,
                "right": 6,
                "top": 6,
              }
            }
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                [
                  {
                    "alignItems": "center",
                    "flexGrow": 1,
                    "justifyContent": "center",
                  },
                  {
                    "borderRadius": 23,
                  },
                ],
              ]
            }
            testID="icon-button"
          >
            <Text />
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "marginTop": 10,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <Image
        source={1}
        style={
          {
            "height": 300,
            "width": 300,
          }
        }
      />
      <View
        style={
          {
            "borderRadius": 6,
            "marginTop": 20,
            "paddingHorizontal": 12,
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            {
              "color": "#1e90ff",
              "fontSize": 20,
              "fontWeight": "700",
            }
          }
        >
          No tasks available
        </Text>
      </View>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#1e90ff",
        "borderRadius": 50,
        "bottom": 20,
        "elevation": 5,
        "opacity": 1,
        "padding": 8,
        "position": "absolute",
        "right": 20,
        "transform": [
          {
            "scale": 1,
          },
        ],
      }
    }
    testID="addButton"
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "transparent",
          "borderRadius": 20,
          "height": 40,
          "margin": 6,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
          "width": 40,
        }
      }
      testID="icon-button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderColor": "rgba(121, 116, 126, 1)",
            "borderRadius": 20,
            "borderWidth": 0,
            "elevation": 0,
            "flex": 1,
            "overflow": "hidden",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="icon-button-container"
      >
        <View
          accessibilityComponentType="button"
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityTraits="button"
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          centered={true}
          collapsable={false}
          color="#fff"
          focusable={true}
          hitSlop={
            {
              "bottom": 6,
              "left": 6,
              "right": 6,
              "top": 6,
            }
          }
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              [
                {
                  "alignItems": "center",
                  "flexGrow": 1,
                  "justifyContent": "center",
                },
                {
                  "borderRadius": 20,
                },
              ],
            ]
          }
          testID="icon-button"
        >
          <Text />
        </View>
      </View>
    </View>
  </View>
</View>
`;
